### ISSUE TITLE
Configuration

### ISSUE THREAD

**Issue by sagikazarmark** [toin0u/concise-cli#1]
I am not sure that phpdotenv is the best solution for this.

I imagine it can work with local and global configuration as well, so that config files can be found at one of the following paths:
- `./concise.{json|yml}`
- `$HOME/concise.{json|yml}`
- `$HOME/.concise.{json|yml}`
- path from `-c` or `--config` console option

Cilex seems to have something similar, but it is removed from the latest develop version, so I guess it will eventually get removed from the latest stable as well. However it does not mean we can't use something similar strategy. Also, Symfony Config is an option as well. I have to look into this to find the best possible solution.

**Comment by sagikazarmark** [comment]
Will try to stick to YAML config

**Comment by toin0u** [comment]
I agree let's use this strategy and the YAML config :+1: 
(I've added to packagist, scutinizer-ci and travis-ci :)

**Comment by sagikazarmark** [comment]
Thanks :smile: I am going to finish the factory thing in concise first, because I want the cli package use it.

**Comment by toin0u** [comment]
I just saw your work on the factory ! I'll give a look asap :)

**Comment by sagikazarmark** [comment]
It is really a basic thing, tied to the HttpAware Providers. I am thinking about leaving it because it might probably end up some sort of dependency responsible class. However the cli needs some sort of factory, so I am not sure about it ATM.

**Comment by toin0u** [comment]
Oh I see - I agree a factory is maybe not needed.
For the CLI, maybe the provider can be given in the config file ?

**Comment by sagikazarmark** [comment]
Providers in general can be any kind of provider. Even one not being a HttpAware adapter. So the CLI probably only have to support our providers, which will probably implement HttpAware. However the Factory should support any kind of providers, this is why I will probably leave it.

The convention is that in case of HttpAware providers the first argument in constructor is the HttpAdapter, any other arguments come after. So it is possible to provide any number of arguments in configuration and instantiate the provider using Reflection (since we are not aware of the number of arguments). They could be only an apiKey, or a username and a password (and an apiKey).

What do you think?

**Comment by toin0u** [comment]
> So the CLI probably only have to support our providers, which will probably implement HttpAware.

I completely agree - let's only support providers which extends `HttpAdapterAware` :)

> However the Factory should support any kind of providers, this is why I will probably leave it.

Make sense :)

> The convention is that in case of HttpAware providers the first argument in constructor is the HttpAdapter, any other arguments come after. So it is possible to provide any number of arguments in configuration and instantiate the provider using Reflection (since we are not aware of the number of arguments). They could be only an apiKey, or a username and a password (and an apiKey).

That sounds like a nice solution - really user friendly :+1:

**Comment by sagikazarmark** [comment]
Great, going that way then

**Comment by toin0u** [comment]
:+1:

**Comment by sagikazarmark** [comment]
I am thinking about supporting multiple instances in one configuration, like this:

``` yml
default:
  provider: google
  arguments:
    apiKey: 1234
secondary:
  provider: google
  arguments:
    apiKey: 4321
```

Or just support single configuration?

`provider` config will also support a class, which must implement `Provider`.

I am also thinking about adding the `Chain` provider as a special on with accepting a set of provider configurations:

``` yml
provider: chain
providers:
  default:
    provider: google
    arguments:
      apiKey: 1234
  secondary:
    provider: google
    arguments:
      apiKey: 4321
```

@toin0u What do you think about all these?

**Comment by sagikazarmark** [comment]
Thinking about validating the configuration using this:
https://github.com/romaricdrigon/MetaYaml

But it might be an overkill.
